### PR TITLE
[feat] Refactor trtllmgen MOE and add Bf16 trtllmgen moe

### PR DESCRIPTION
**Pull request by jiahanc**
## 📌 Description
- Refactor `trtllm_fused_moe_kernel_launcher.cu` to use class structure for code cleanliness and readability
- Add BF16 MOE, initial PR (https://github.com/flashinfer-ai/flashinfer/pull/1859) from @aleozlx and @nekorobov
- Add BF16 MOE autotune

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * BF16 Mixture-of-Experts (MoE) pathway added with autotuning and public API access.

* **Improvements**
  * Unified BF16/FP8/FP4/FP16 pathways with clearer dtype compatibility checks and corrected operator return semantics.
  * Routing selection now respects token-size and input packing, and diagnostics produce more descriptive error messages.

* **Tests**
  * Expanded BF16 test coverage across routing modes, weight layouts, and token sizes.

* **Chores**
  * Updated artifact metadata and checksums.

### CODE REVIEW

File: csrc/trtllm_fused_moe_kernel_launcher.cu [hunk 1]
```diff
@@ -74,1095 +76,1426 @@ std::set<int32_t> computeSelectedTileN(std::vector<int32_t> const& supported_til
...
+    TensorView gemm1_weights, TensorView output1_scales_scalar,
+    TensorView output1_scales_gate_scalar, TensorView gemm2_weights,
+    TensorView output2_scales_scalar, TensorView output, int64_t num_experts, int64_t top_k,
+    Optional<int64_t> n_group, Optional<int64_t> topk_group, int64_t intermediate_size,
+    int64_t local_expert_offset, int64_t local_num_experts, Optional<double> routed_scaling_factor,
+    bool use_routing_scales_on_input, int64_t routing_method_type, bool enable_pdl,
+    Array<int64_t> config_index) {
+  // Basic type validation
+  auto dtype = hidden_states.dtype();
+  TVM_FFI_ICHECK(dtype == dl_float8_e4m3fn || dtype == dl_float16 || dtype == dl_bfloat16)
```
Comment (nekorobov) [inline csrc/trtllm_fused_moe_kernel_launcher.cu:1338]: Check routing logits and routing bias here.

File: csrc/trtllm_fused_moe_kernel_launcher.cu [hunk 2]
```diff
@@ -74,1095 +76,1426 @@ std::set<int32_t> computeSelectedTileN(std::vector<int32_t> const& supported_til
...
+    TensorView routing_logits, Optional<TensorView> routing_bias, TensorView hidden_states,
+    TensorView hidden_states_scale, TensorView gemm1_weights, TensorView gemm1_weights_scale,
+    TensorView gemm2_weights, TensorView gemm2_weights_scale, TensorView output,
+    int64_t num_experts, int64_t top_k, Optional<int64_t> n_group, Optional<int64_t> topk_group,
+    int64_t intermediate_size, int64_t local_expert_offset, int64_t local_num_experts,
+    Optional<double> routed_scaling_factor, int64_t routing_method_type, bool use_shuffled_weight,
+    int64_t weight_layout, bool enable_pdl, Array<int64_t> config_index) {
+  // Basic type validation
+  auto dtype = hidden_states.dtype();
+  TVM_FFI_ICHECK(dtype == dl_float16 || dtype == dl_bfloat16 || dtype == dl_float8_e4m3fn)
```
Comment (nekorobov) [inline csrc/trtllm_fused_moe_kernel_launcher.cu:1427]: Here as well

File: csrc/trtllm_fused_moe_kernel_launcher.cu (hunk 1 above)
Comment (jiahanc) [reply csrc/trtllm_fused_moe_kernel_launcher.cu:1338]: The checks are in the launcher class, but agree moving there is more reasonable 😄

**Review by IwakuraRein — APPROVED** [review]
Thanks for your contributions!

File: flashinfer/fused_moe/core.py [hunk 3]
```diff
@@ -177,6 +178,39 @@ class GatedActType(IntEnum):
     GeGlu = 1
 
 
+def is_flashinfer_trtllm_moe_supported(
```
Comment (yzh119) [inline flashinfer/fused_moe/core.py]: Can we cache the value with `functools.cache`?

Comment (yzh119) [reply flashinfer/fused_moe/core.py]: Regarding the naming, I prefer `is_trtllm_moe_supported`, no need to mention flashinfer here.

File: tests/moe/test_trtllm_gen_fused_moe.py [hunk 4]
```diff
@@ -2102,16 +2285,16 @@ def run_moe_test(
     [
         pytest.param(
             {
-                "num_experts": 256,
+                "num_experts": 128,
```
Comment (yzh119) [inline tests/moe/test_trtllm_gen_fused_moe.py:2288]: Why changing this value? Can we test both?

Comment (jiahanc) [reply tests/moe/test_trtllm_gen_fused_moe.py:2288]: Just want to speed up testing and there is no model use this config (256 expert), will add both config for better coverage

**Review by yzh119 — APPROVED** [review]
LGTM
